File: 2016/12.py
from Utils import *

# Part 1

inp = rdl("12.txt")

reg = {'a': 0,
       'b': 0,
       'c': 1,
       'd': 0}

i = 0
while i < len(inp):
    line = inp[i]
    ss = line.split(' ')

    if ss[0] == 'cpy':
        if ss[1].isnumeric():
            reg[ss[2]] = int(ss[1])
        else:
            reg[ss[2]] = reg[ss[1]]
        i += 1
    elif ss[0] == 'inc':
        reg[ss[1]] += 1
        i += 1
    elif ss[0] == 'dec':
        reg[ss[1]] -= 1
        i += 1
    elif ss[0] == 'jnz':
        is_zero = False
        if ss[1].isnumeric():
            is_zero = int(ss[1]) == 0
        else:
            is_zero = reg[ss[1]] == 0

        if not is_zero:
            i += int(ss[2])
        else:
            i += 1

print(reg)

# 3:19 (1st)

# Part 2

# 3:43 (1st)

File: v2/2016/12.R

# Load --------------------------------------------------------------------

rm(list = ls())
source("utilities.R")

# Problem -----------------------------------------------------------------

# inp <- strsplit("cpy 41 a
# inc a
# inc a
# dec a
# jnz a 2
# dec a", "\n")[[1]]  # Example

inp <- get_input("2016/12", parse = F, user = "blakeebets", cache = T)

reg <- setNames(rep(0, 4), letters[1:4])
reg[["c"]] <- 1

i <- 1
last_d <- reg[["d"]]
while(i <= length(inp)) {
    ls <- strsplit(inp[i], " ")[[1]]
    # print(ls)
    
    if(ls[1] == "cpy") {
        source <- ls[2]
        x <- if(source %in% names(reg)) reg[[source]] else as.numeric(source)
        
        reg[[ls[3]]] <- x
        i <- i + 1
    } else if(ls[1] == "inc") {
        reg[[ls[2]]] <- reg[[ls[2]]] + 1
        i <- i + 1
    } else if(ls[1] == "dec") {
        reg[[ls[2]]] <- reg[[ls[2]]] - 1
        i <- i + 1
    } else if(ls[1] == "jnz") {
        source <- ls[2]
        x <- if(source %in% names(reg)) reg[[source]] else as.numeric(source)
        
        if(x != 0) {
            source <- ls[3]
            y <- if(source %in% names(reg)) reg[[source]] else as.numeric(source)
            i <- i + y
        } else {
            i <- i + 1
        }
    }
    
    if(reg[["d"]] != last_d) {
        print(reg)
        last_d <- reg[["d"]]
    }
}

# 6:51 (5th)

# 11:44 (40th)

